### PR TITLE
[trust] Add Yoran-Oran gambits

### PR DESCRIPTION
**Pull request by thecursedcometh2**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

This PR adds most of the proper skills and conditions for Yoran-Oran (UC), based on wiki information and my own observations when using him. His MP currently does not scale with Unity ranking, nor is he able to use the Nott weaponskill. 

## Steps to test these changes

- Summon Yoran-Oran (UC).
- Observe new behaviors.

### CODE REVIEW

**Review by Xaver-DaRed — APPROVED** [review]
Looks good to me at first glance.

**Comment by zach2good** [comment]
```
CREATE TABLE `mob_pools` (
  `poolid` int(10) unsigned NOT NULL,
  `name` varchar(24) DEFAULT NULL,
  `packet_name` varchar(24) DEFAULT NULL,
  `familyid` smallint(4) unsigned NOT NULL DEFAULT 0,
  `modelid` binary(20) NOT NULL,
  `mJob` tinyint(2) unsigned NOT NULL DEFAULT 1,
  `sJob` tinyint(2) unsigned NOT NULL DEFAULT 1,
  `cmbSkill` tinyint(2) unsigned NOT NULL DEFAULT 1,
  `cmbDelay` smallint(3) unsigned NOT NULL DEFAULT 240,
  `cmbDmgMult` smallint(4) unsigned NOT NULL DEFAULT 100,
  `behavior` smallint(5) unsigned NOT NULL DEFAULT 0,
  `aggro` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `true_detection` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `links` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `mobType` smallint(5) unsigned NOT NULL DEFAULT 0,
  `immunity` int(10) NOT NULL DEFAULT 0,
  `name_prefix` tinyint(4) unsigned NOT NULL DEFAULT 0,
  `flag` int(11) unsigned NOT NULL DEFAULT 0,
  `entityFlags` int(11) unsigned NOT NULL DEFAULT 0,
  `animationsub` tinyint(1) NOT NULL DEFAULT 0,
  `hasSpellScript` tinyint(1) unsigned NOT NULL DEFAULT 0, -- NO LONGER IN USE. DOES NOTHING. TODO: Remove me
  `spellList` smallint(4) NOT NULL DEFAULT 0,
  `namevis` tinyint(4) NOT NULL DEFAULT 1,
  `roamflag` smallint(3) unsigned NOT NULL DEFAULT 0,
  `skill_list_id` smallint(5) unsigned NOT NULL DEFAULT 0,
  `resist_id` smallint(5) unsigned NOT NULL DEFAULT 0,
  PRIMARY KEY (`poolid`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```

```
INSERT INTO `mob_pools` VALUES (5980,'yoran-oran_uc','Yoran-Oran',153,0x0000170C00000000000000000000000000000000,3,0,3,240,100,0,0,0,0,0,0,32,0,3,0,0,393,0,0,1095,153);
```

https://github.com/LandSandBoat/server/blob/d2a9dc48f3916bded3c767aba1f511012d0c1d82/src/map/utils/trustutils.cpp#L412

https://github.com/LandSandBoat/server/blob/base/src/map/ai/controllers/trust_controller.cpp#L137

```cpp
// PTrust->m_MovementType is read from 'behaviour' in a trust's mob_pool entry
enum TRUST_MOVEMENT_TYPE
{
    MELEE_RANGE        = 0, // Default
    NO_MOVE            = 1,
    MID_RANGE          = 2,
    LONG_RANGE         = 3,
    LAST_MOVEMENT_TYPE = 4,
};
```
```
SELECT name, behavior FROM mob_pools WHERE name LIKE "%Yoran%";

>>> yoran-oran_uc, 0
```

Yoran doesn't live at melee range, that's one of his defining and most important features. He needs his movement set to LONG_RANGE

**Comment by thecursedcometh2** [comment]
I was under the impression that the movement types weren't implemented. With that knowledge I'm going to add movement types for some other trusts (in another PR obviously).

**Comment by thecursedcometh2** [comment]
Sorry, I didn't realize discarding the previous commits would close the PR. Re-opening with changes from #4032 integrated.

File: scripts/globals/spells/trust/yoran-oran_uc.lua [hunk 1]
```diff
@@ -16,6 +17,31 @@ end
...
+    -- TODO: UC trusts are supposed to get bonuses depending on unity ranking. Needs research.
+    -- TODO: Custom spawn messages if Unity ranking is higher.
+    -- TODO: Figure out what level of Fast Cast he has.
+
+    mob:addMod(xi.mod.CURE_POTENCY, 50)
+    mob:addMod(xi.mod.REGAIN, 50)
+    mob:addMod(xi.mod.MPP, 15) -- This is supposed to increase with Unity rank, but I don't believe that's implemented so it is set to the minimum.
+    mob:addSimpleGambit(ai.t.SELF, ai.c.NOT_STATUS, xi.effect.AFFLATUS_SOLACE, ai.r.JA, ai.s.SPECIFIC, xi.ja.AFFLATUS_SOLACE)
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.HPP_LT, 25, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.CURE)
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.HPP_LT, 75, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.CURE)
```
Comment (zach2good) [inline scripts/globals/spells/trust/yoran-oran_uc.lua]: I'd probably leave a note here mentioning that this _isn't_ redundant, this is giving Yoran more chances to be able to heal if someone in the party is at or under 25 HP (and therefore also under 75)

File: scripts/globals/spells/trust/yoran-oran_uc.lua [hunk 2]
```diff
@@ -16,6 +17,31 @@ end
...
     xi.trust.message(mob, xi.trust.messageOffset.SPAWN)
 
+    -- TODO: Nott weaponskill needs implemented and logic added here for Yoran-Oran to use at 50% MP at level 50.
+    -- TODO: UC trusts are supposed to get bonuses depending on unity ranking. Needs research.
+    -- TODO: Custom spawn messages if Unity ranking is higher.
+    -- TODO: Figure out what level of Fast Cast he has.
+
+    mob:addMod(xi.mod.CURE_POTENCY, 50)
+    mob:addMod(xi.mod.REGAIN, 50)
+    mob:addMod(xi.mod.MPP, 15) -- This is supposed to increase with Unity rank, but I don't believe that's implemented so it is set to the minimum.
```
Comment (zach2good) [inline scripts/globals/spells/trust/yoran-oran_uc.lua]: Can you put a newline in here to make the mods separate from the gambits?

Comment (zach2good) [inline scripts/globals/spells/trust/yoran-oran_uc.lua]: This comment looks like a TODO, and should be marked as such

File: scripts/globals/spells/trust/yoran-oran_uc.lua [hunk 3]
```diff
@@ -16,6 +17,31 @@ end
...
+    mob:addSimpleGambit(ai.t.SELF, ai.c.NOT_STATUS, xi.effect.AFFLATUS_SOLACE, ai.r.JA, ai.s.SPECIFIC, xi.ja.AFFLATUS_SOLACE)
```
Comment (zach2good) [inline scripts/globals/spells/trust/yoran-oran_uc.lua:29]: Does AFFLATUS_SOLACE work with Yoran here? Is he leaving behind stoneskin on people he heals, etc?

File: scripts/globals/spells/trust/yoran-oran_uc.lua [hunk 4]
```diff
@@ -16,6 +17,31 @@ end
...
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.HPP_LT, 25, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.CURE)
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.HPP_LT, 75, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.CURE)
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.NOT_STATUS, xi.effect.PROTECT, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.PROTECTRA)
+    mob:addSimpleGambit(ai.t.PARTY, ai.c.NOT_STATUS, xi.effect.SHELL, ai.r.MA, ai.s.HIGHEST, xi.magic.spellFamily.SHELLRA)
```
Comment (zach2good) [inline scripts/globals/spells/trust/yoran-oran_uc.lua:33]: There -ra's probably want to be PARTY_MULTI(https://github.com/LandSandBoat/server/blob/base/scripts/globals/gambits.lua#L23)

**Review by zach2good — CHANGES_REQUESTED** [review]
Looking close to ready 👍 

Can you make your commit message something more descriptive? That's the message that gets left in the codebase forever and needs to be representative of what changes you're making. `Added changes from #4032` doesn't convey what's changing here.

File: scripts/globals/spells/trust/yoran-oran_uc.lua (hunk 3 above)
Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua:29]: Yes, he is.

File: scripts/globals/spells/trust/yoran-oran_uc.lua (hunk 2 above)
Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua]: Done.

Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua]: Done.

File: scripts/globals/spells/trust/yoran-oran_uc.lua (hunk 1 above)
Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua]: Done.

File: scripts/globals/spells/trust/yoran-oran_uc.lua (hunk 4 above)
Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua:33]: Done.

Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua:33]: Hmm.... he's stopped casting Protectra/Shellra under the PARTY_MULTI condition. Should I revert it or is there something else going on? No errors in the log.

Comment (zach2good) [reply scripts/globals/spells/trust/yoran-oran_uc.lua:33]: You can check the conditions and debug/see:
https://github.com/LandSandBoat/server/blob/base/src/map/ai/helpers/gambits_container.cpp#L229

Presumably the predicate that picks a target to cast on isn't actually doing that, or it isn't handled correctly:
https://github.com/LandSandBoat/server/blob/base/src/map/ai/helpers/gambits_container.cpp#L290

For the sake of getting this PR in, just set it back to PARTY

Comment (thecursedcometh2) [reply scripts/globals/spells/trust/yoran-oran_uc.lua:33]: Done. I'll look at it in my own time.

**Review by zach2good — APPROVED** [review]

**Pull request merged**